src/modifierManaBarrier.ts

```diff
@@ -0,0 +1,101 @@
+    add: (unit: Unit.IUnit, underworld: Underworld, prediction: boolean, quantity: number = 1) => {
+      getOrInitModifier(unit, manaBarrierId, { isCurse: false, quantity, keepOnDeath: true }, () => {
+        Unit.addEvent(unit, manaBarrierId);
+
+        // Add subsprite image
+        const animatedBarrierSprite = Image.addSubSprite(unit.image, modifierImagePath);
+        if (animatedBarrierSprite) {
+          // Make it blue just so it looks distinct from shield
+          animatedBarrierSprite.tint = 0x0000ff;
+        }
+      });
+
+      if (!prediction) {
+        updateTooltip(unit);
+      }
+    },
+    subsprite: {
+      imageName: modifierImagePath,
+      alpha: 0.5,
+      anchor: {
+        x: 0.5,
+        y: 0.5,
+      },
+      scale: {
+        x: 1.0,
+        y: 1.0,
+      },
+    },
+  });
+  registerEvents(manaBarrierId, {
+    onTakeDamage: (unit: Unit.IUnit, amount: number, underworld: Underworld, prediction: boolean, damageDealer?: Unit.IUnit) => {
+      const modifier = unit.modifiers[manaBarrierId];
+      if (modifier) {
+        // If the incoming effect is damage (not healing)
+        if (amount > 0) {
+          // Damage should be taken from mana before health
+          const blockableDamage = Math.floor(Math.min(amount, unit.mana * CalcMult(modifier.quantity)));
+          if (blockableDamage > 0) {
+            amount -= blockableDamage;
+            unit.mana -= blockableDamage / CalcMult(modifier.quantity);
+            unit.mana = Math.floor(unit.mana); // To prevents rounding errors
+            if (!prediction) {
+              floatingText({
+                coords: unit,
+                text: 'Shielded from damage!',
+                style: {
+                  fill: 'blue',
+                  ...config.PIXI_TEXT_DROP_SHADOW
+                },
+              });
+            }
+            updateTooltip(unit);
+          }
+        }
+      }
+
+      // Does not modify incoming damage/healing otherwise
+      return amount;
+    },
+    onTurnStart: async (unit: Unit.IUnit, underworld: Underworld, prediction: boolean) => {
+      const modifier = unit.modifiers[manaBarrierId];
+      if (modifier) {
+        updateTooltip(unit);
+      }
+    }
+  });
+}
+
+export function updateTooltip(unit: Unit.IUnit) {
+  const modifier = unit.modifiers[manaBarrierId];
+  if (modifier) {
+    // Set tooltip:
+    // modifier.tooltip = `${modifier.quantity}% ${i18n('Mana Barrier')}`
+
+    // This returns the maximum blockable damage,
+    // based on the unit's mana and modifier quantity
+    // We need to update this any time mana changes
+    modifier.tooltip = `${Math.floor(unit.mana * CalcMult(modifier.quantity))} ${i18n('Mana Barrier')}`
+  }
+}
+
+function CalcMult(quantity: number): number {
+  // Each Quantity = 1% barrier effectiveness
+  // 100(%) quantity = 1:1 ratio = 1 mana blocks 1 damage
+  // 20(%) quantity = 5:1 ratio = 5 mana blocks 1 damage
+  return quantity / 100;
+}
```

Comment:
Why not 1 quantity = 10% or something.  Why out of 100?

Reply:
This way it won't have to be upgraded so many times.  I know we discussed this the other day, but rather than making pushing the "plus button" do 5 quantity we should just change the runes to be more effective per upgrade IMO.  Thoughts?

Reply:
This is for the same reason shield/poison now give 1 shield/poison per quantity instead of 30/20. Being able to add modifiers in smaller quantities is what gives Overheal, Shield Regen, and OnHitPoison room to exist.

Debilitate may be a closer example: If we have a rune/spell/enemy that applies some amount of debilitate when it hits, we may not want to apply a full 100% increased damage every time. One source may inflict 5% while another inflicts 50%, and having 1 quantity = 1% makes sense in this case. It also lets us easily change any source without affecting another one. 

Technically, 1 quantity only ever needs to be = the lowest common denominator of each modifier: If we always apply debilitate in multiples of 10%, we _could_ make 1 quantity of debilitate = 10%, but what do we gain by doing that? Giving every modifier a different quantity-to-% ratio is confusing, and the only benefit I can see is adding modifiers through the power bar more quickly, which doesn't outweigh the balance and design limitations imo. For runes, a "quantity per upgrade" variable seems necessary either way. So I think it makes sense to have these % based modifiers all use the same 1:1% ratio. I can read an effect and know immediately that 10 quantity = 10% without having to check the modifier logic. If I want to add a 5% source, I don't have to change the base 1:10% ratio of the modifier and double the quantity applied by all other sources

Reply:
Gotcha, okay that works for me.